x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/normalize/NormalizePipelineNormalizer.java

```diff
@@ -0,0 +1,140 @@
+
+    static class RescaleZeroToOne extends SinglePassSimpleStatisticsNormalizer {
+        static final String NAME = "rescale_0_1";
+
+        RescaleZeroToOne(List<Double> values) {
+            super(values);
+        }
+
+        @Override
+        double normalize(double value) {
+            return (value - min) / (max - min);
+        }
+    }
+
+    static class RescaleZeroToOneHundred extends SinglePassSimpleStatisticsNormalizer {
+        static final String NAME = "rescale_0_100";
+
+        RescaleZeroToOneHundred(List<Double> values) {
+            super(values);
+        }
+
+        @Override
+        double normalize(double value) {
+            return 100 * (value - min) / (max - min);
+        }
+    }
+
+    static class Mean extends SinglePassSimpleStatisticsNormalizer {
+        static final String NAME = "mean";
+
+        Mean(List<Double> values) {
+            super(values);
+        }
+
+        @Override
+        double normalize(double value) {
+            return (value - mean) / (max - min);
+        }
+    }
+
+    static class Percent extends SinglePassSimpleStatisticsNormalizer {
+        static final String NAME = "percent";
+
+        Percent(List<Double> values) {
+            super(values);
+        }
+
+        @Override
+        double normalize(double value) {
+            return value / sum;
+        }
+    }
+
+    static class ZScore extends SinglePassSimpleStatisticsNormalizer {
+        static final String NAME = "z-score";
+
+        private final double stdev;
+
+        ZScore(List<Double> values) {
+            super(values);
+            double variance = 0.0;
+            for (Double value : values) {
+                if (value.isNaN() == false) {
+                    variance += Math.pow(value - mean, 2);
+                }
+            }
+            this.stdev = Math.sqrt(variance / count);
+        }
+
+        @Override
+        double normalize(double value) {
+            return (value - mean) / stdev;
+        }
+    }
+
+    static class Softmax extends NormalizePipelineNormalizer {
+        static final String NAME = "softmax";
+
+        private double sumExp;
+
+        Softmax(List<Double> values) {
+            double sumExp = 0.0;
+            for (Double value :  values) {
+                if (value.isNaN() == false) {
+                    sumExp += Math.exp(value);
+                }
+            }
+
+            this.sumExp = sumExp;
+        }
+
+        @Override
+        double normalize(double value) {
+            return Math.exp(value) / sumExp;
+        }
+    }
+
+    abstract double normalize(double value);
+
+    abstract static class SinglePassSimpleStatisticsNormalizer extends NormalizePipelineNormalizer {
+        protected final double max;
+        protected final double min;
+        protected final double sum;
+        protected final double mean;
+        protected final int count;
+
+        SinglePassSimpleStatisticsNormalizer(List<Double> values) {
```

Comment:
Could this take a `DoubleStream` instead?

Reply:
You'd only get a single pass but you wouldn't need to make `Double` objects which is kind of nice.

Reply:
Not that pipeline aggs are all that efficient here, but I feel compelled to save autoboxing when I can.